rust/core-lib/src/file.rs

```diff
@@ -186,7 +186,15 @@ where
 }
 
 fn try_save(path: &Path, text: &Rope, encoding: CharacterEncoding) -> io::Result<()> {
-    let mut f = File::create(path)?;
+    use std::fs;
+
+    let tmp_extension = path
+        .extension()
+        .and_then(|extension| extension.to_str()?.to_string().into())
+        .map_or_else(|| "swp".to_string(), |extension| extension + ".swp");
+    let tmp_path = &path.with_extension(tmp_extension);
+
+    let mut f = File::create(tmp_path)?;
```

Comment:
This will fail of the swap file already exists. We probably don't want that. What do we want? Is this a situation where we might consider alerting the user?

Reply:
It doesn't fail if the swap file exists (according to [docs](https://doc.rust-lang.org/src/std/fs.rs.html#370) and my light testing with `xi-gtk`), but I agree that the behavior of ignoring and overwriting the swapfile is a bit weird.

I'm thinking maybe instead of saying "hey, you're going to overwrite the temp file" when trying to save, we could give the user the option to restore from the swapfile if it's still hanging around when they reopen the file? It seems to make more sense that way since we're adding this to prevent data-loss?

**However**, this seems to have a suspiciously high level of overlap with https://github.com/xi-editor/xi-editor/pull/939 I think. Maybe this Atomic Save is just a poor man's version of the `Session`object from the RFC? If that's the case, it's quite possible that this PR isn't really a super useful addition in the long run.

This is more of an abstract design question rather than an implementation detail; maybe it warrants more discussion to properly flesh out.

Reply:
Ah okay I'd misread the docs, this should overwrite the file.

I agree that there's really no great way for us to surface this information to the user, right now: ideally we would want to be able to open the temp file for them, and _really ideally_ we'd want to be able to show them a diff; but we don't really have either of these functions atm.

So I'll walk my suggestion back: I think that the overwrite is fine for now, and it's a nice theoretical win that if we screw up while writing we at least don't corrupt the existing file.